### PR TITLE
Add i.MX93 M33 DDR code support

### PR DESCRIPTION
**Pull request by alxlastur**
This commit adds the support for running the Cortex M33 code from DDR memory.

### CODE REVIEW

File: boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_CLOCK_CONTROL=y
+CONFIG_PINCTRL=y
+CONFIG_SERIAL=y
```
Comment (dbaluta) [inline boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig:6]: Remove CONFIG_PINCTRL from boards config. 

See 82a192c8a92b66a2b commit for example.

**Review by iuliana-prodan — DISMISSED** [review]

File: boards/nxp/imx93_evk/doc/index.rst [hunk 2]
```diff
@@ -221,10 +221,20 @@ prompt.
 
 Use U-Boot to load and kick zephyr.bin to Cortex-M33 Core:
 
+CODE TCM
+========
```
Comment (dbaluta) [inline boards/nxp/imx93_evk/doc/index.rst:225]: I would say here "Boot with code from TCM"

File: boards/nxp/imx93_evk/doc/index.rst [hunk 3]
```diff
@@ -221,10 +221,20 @@ prompt.
...
+
 .. code-block:: console
 
     load mmc 1:1 0x80000000 zephyr.bin;cp.b 0x80000000 0x201e0000 0x30000;bootaux 0x1ffe0000 0
 
+DDR
+===
```
Comment (dbaluta) [inline boards/nxp/imx93_evk/doc/index.rst:232]: ```Boot with code from DDR```

File: boards/nxp/imx93_evk/doc/index.rst (hunk 2 above)
Comment (alxlastur) [reply boards/nxp/imx93_evk/doc/index.rst]: done

File: boards/nxp/imx93_evk/doc/index.rst (hunk 3 above)
Comment (alxlastur) [reply boards/nxp/imx93_evk/doc/index.rst]: done

File: boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts [hunk 4]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+
+#include <nxp/nxp_imx93_m33.dtsi>
```
Comment (LaurentiuM1234) [inline boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts:9]: Why not just include `imx93_evk_mimx9352_m33.dts` here?

File: boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_CLOCK_CONTROL=y
+CONFIG_SERIAL=y
+CONFIG_UART_CONSOLE=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_CONSOLE=y
+CONFIG_XIP=y
```
Comment (LaurentiuM1234) [inline boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig:9]: What's this for? Isn't the memory supposed to be RW? Also, why the code/data split?

File: dts/arm/nxp/nxp_imx93_m33.dtsi [hunk 6]
```diff
@@ -40,6 +40,18 @@
 			reg = <0x20000000 DT_SIZE_K(128)>;
 		};
 
+		ddr_code: code@84000000 {
```
Comment (LaurentiuM1234) [inline dts/arm/nxp/nxp_imx93_m33.dtsi:43]: Isn't this board-related?

File: boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts (hunk 4 above)
Comment (alxlastur) [reply boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts:9]: I followed the structure from imx8mp m7 ddr support: https://github.com/zephyrproject-rtos/zephyr/blob/a2ac676a106a757fe4cb2664bab35ea985da95f9/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_m7_ddr.dts#L9

Should I use the structure from the imx95_evk?

Comment (LaurentiuM1234) [reply boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts:9]: sure, if it makes more sense

File: boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig (hunk 5 above)
Comment (alxlastur) [reply boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig:9]: I followed the structure from imx8mp m7 ddr support:
https://github.com/zephyrproject-rtos/zephyr/blob/2f23313e37e431d64972133e3c946a2829dad142/boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_m7_ddr_defconfig#L12

I tested for option CONFIG_XIP=y. Should we use CONFIG_XIP=n? As I understand from the https://github.com/zephyrproject-rtos/zephyr/blob/2f23313e37e431d64972133e3c946a2829dad142/include/zephyr/arch/arm/cortex_m/scripts/linker.ld#L22 the CONFIG_XIP=y creates the code/data split. That's why it's necessary to zephyr,flash and zephyr,sram in [boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr.dts](https://github.com/zephyrproject-rtos/zephyr/pull/79993/files/3d1349c4c1df78d4c2f777f60dffe7f979f913bb#diff-7d0ac3cd20b2794ee8e8c79f9404c7b838918b427138caba6143900ce56c1d4f)

Is CONFIG_XIP=y not the right way ?

Comment (LaurentiuM1234) [reply boards/nxp/imx93_evk/imx93_evk_mimx9352_m33_ddr_defconfig:9]: I'd suggest disabling it, don't see a reason why we'd want to use it with everything placed in DDR

**Review by JiafeiPan — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

File: doc/kernel/data_structures/index.rst [hunk 7]
```diff
@@ -20,10 +20,12 @@ behind this design is to allow the collections to be used in contexts
 where dynamic allocation is disallowed (i.e. there is no need to
 allocate node objects because the memory is provided by the user).
 
-Note also that these libraries are uniformly unsynchronized; access to
+Note also that these libraries are generally unsynchronized; access to
```
Comment (LaurentiuM1234) [inline doc/kernel/data_structures/index.rst]: Wasn't this already merged via https://github.com/zephyrproject-rtos/zephyr/pull/79943?

**Comment by dbaluta** [comment]
@alxlastur  second patch from this PR doesn't belong here. Please remove it.

https://github.com/zephyrproject-rtos/zephyr/pull/79993/commits/aa6f3fb4961d2ed3d7d24a610e06d63f4c61b0da

**Comment by alxlastur** [comment]
Hi @JiafeiPan @dbaluta @fabiobaltieri 
I made the rebase and removed the extra commit.
Can you please reapprove?

Thank you

**Review by LaurentiuM1234 — DISMISSED** [review]

**Review by JiafeiPan — DISMISSED** [review]

**Review by iuliana-prodan — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

**Comment by JiafeiPan** [comment]
> @JiafeiPan any idea why twister check fails. It seems unrelated to this PR.

Yes, you are correct, not sure why this issue is not found before, I have created PR to fix this twister issue: 
#82586 

so let's waiting the fix is merged then merge this one after all checking pass.

**Comment by JiafeiPan** [comment]
Twister issue has been fixed, please rebase and trigger twister again,

**Review by iuliana-prodan — APPROVED** [review]

**Review by dbaluta — APPROVED** [review]

**Review by LaurentiuM1234 — APPROVED** [review]

**Review by JiafeiPan — APPROVED** [review]

**Pull request merged**